src/rcdom.rs

```diff
@@ -244,15 +245,14 @@ impl TreeSink for RcDom {
 
     fn add_attrs_if_missing(&mut self, target: Handle, mut attrs: Vec<Attribute>) {
         let mut node = target.borrow_mut();
-        // FIXME: mozilla/rust#15609
-        let existing = match node.deref_mut().node {
-            Element(_, ref mut attrs) => attrs,
-            _ => return,
+        let existing = if let Element(_, ref mut attrs) = node.deref_mut().node {
+            attrs
+        } else {
+            panic!("not an element")
         };
 
-        // FIXME: quadratic time
-        attrs.retain(|attr|
-            !existing.iter().any(|e| e.name == attr.name));
+        let names = attrs.iter().map(|e| e.name.clone()).collect::<HashSet<_>>();
```

Comment:
I fixed that, but forgot to commit before pushing… Will address the other points while at it.